firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Simulator.cpp
+ *
+ *  Created on: May 30, 2021
+ *      Author: Rohan
+ *
+ *  Simulator for power management
+ */
+
+#include "Pufferfish/Driver/Power/Simulator.h"
+
+#include <cmath>
+#include <random>
+
+#include "Pufferfish/Util/Timeouts.h"
+
+namespace Pufferfish::Driver::Power {
+
+// NOLINTNEXTLINE(cert-msc51-cpp)
+std::ranlux24_base prng;
+// NOLINTNEXTLINE(readability-magic-numbers)
+std::uniform_int_distribution<int> uniform_int(1, 100);
+static constexpr float uniform_width = 1;
+std::uniform_real_distribution<float> uniform_centered(-uniform_width / 2, uniform_width / 2);
+
+void Simulator::input_clock(uint32_t current_time) {
+  if (initial_time_ == 0) {
+    initial_time_ = current_time;
+  }
+  if (update_needed()) {
+    previous_time_ = current_time_;
+  }
+  current_time_ = current_time - initial_time_;
+}
+
+void Simulator::transform(uint32_t current_time, PowerManagement &power_management) {
+  input_clock(current_time);
```

Comment:
When you call `input_clock` here and it determines that an update is needed, then it sets `previous_time_ = current_time_`. Next, you call `transform_charge`/`transform_discharge`, which does nothing if `previous_time_ == current_time_` (as determined by an `update_needed()` check). I haven't tried running the code, but this code seems suspicious to me.

Reply:
It looks like this same pattern is used in the BreathingCircuit's simulator, so that needs further investigation.

Reply:
Looks like I missed a basic thing which explains all of this: right after `previous_time` is set to `current_time_`, `current_time_` is changed to a new value.